File: web/src/lib/dropText.ts
// SPDX-FileCopyrightText: 2026 Angels' Ware
// SPDX-License-Identifier: AGPL-3.0-or-later

import { primaryLanguage } from "./language";
import type { Manifest } from "../types";

type LabelledDrop = Pick<Manifest, "label"> &
  Partial<Pick<Manifest, "id" | "subtitle" | "description">>;

// descriptionFingerprint is a 32-bit FNV-1a over Unicode code points, hex,
// zero-padded to 8 chars. Descriptions are paragraphs, so unlike the short
// strings above they are keyed by DROP ID rather than by their English text —
// duplicating 53k characters of prose to use it as a key would be unreadable
// and whitespace-fragile. The fingerprint restores what natural keys gave for
// free: each translation records the fingerprint of the English it was made
// from, and a paragraph edited on the Go side no longer matches, so the reader
// falls back to the new English instead of reading a stale Swedish paragraph.
//
// Mirror in any language (this is how the recorded values were produced):
//   h = 2166136261
//   for cp in text:  h = ((h ^ cp) * 16777619) & 0xFFFFFFFF
//   "%08x" % h
export function descriptionFingerprint(text: string): string {
  let h = 0x811c9dc5;
  for (const ch of text) {
    h = Math.imul(h ^ (ch.codePointAt(0) as number), 0x01000193) >>> 0;
  }
  return h.toString(16).padStart(8, "0");
}

const EN_CATEGORIES: Record<string, string> = {
  ai: "AI",
  flow_control: "Flow control",
  io: "Files & data",
  logic: "Logic",
  network: "Apps & services",
  system: "System",
  transformation: "Change data",
  trigger: "Triggers",
};

export type DescriptionMap = Record<string, { en: string; sv: string }>;

export type Vocabulary = {
  labels: Record<string, string>;
  subtitles: Record<string, string>;
  descriptions: DescriptionMap;
  categories: Record<string, string>;
  ports: Record<string, string>;
  fieldTitles: Record<string, string>;
  fieldHelp: Record<string, string>;
  enums: Record<string, string>;
  connections: Record<string, string>;
  nodeState: Record<string, string>;
  prose: DescriptionMap;
  appNames: Record<string, string>;
};

const VOCABULARY: Record<string, Vocabulary> = {};

const VOCABULARY_LOADERS: Record<string, () => Promise<Vocabulary>> = {
  sv: () => import("../i18n/drops/sv").then((m) => m.SV_VOCABULARY),
};

export async function loadVocabulary(lang: string | undefined): Promise<void> {
  const code = primaryLanguage(lang);
  if (!code || VOCABULARY[code]) return;
  const load = VOCABULARY_LOADERS[code];
  if (!load) return;
  try {
    VOCABULARY[code] = await load();
  } catch {
    /* untranslated beats blocked: resolvers keep returning English */
  }
}

export function registerVocabulary(lang: string, v: Vocabulary): void {
  VOCABULARY[primaryLanguage(lang)] = v;
}

function vocabularyFor(lang: string | undefined): Vocabulary | undefined {
  if (!lang) return undefined;
  return VOCABULARY[primaryLanguage(lang)];
}

export function dropLabel(drop: LabelledDrop, lang?: string): string {
  const v = vocabularyFor(lang);
  return v?.labels[drop.label] ?? drop.label;
}

export function dropSubtitle(drop: LabelledDrop, lang?: string): string {
  const sub = drop.subtitle ?? "";
  if (!sub) return "";
  const v = vocabularyFor(lang);
  return v?.subtitles[sub] ?? sub;
}

export function dropDescription(drop: LabelledDrop, lang?: string): string {
  const desc = drop.description ?? "";
  if (!desc || !drop.id) return desc;
  const v = vocabularyFor(lang);
  const entry = v?.descriptions[drop.id];
  if (!entry) return desc;
  return entry.en === descriptionFingerprint(desc) ? entry.sv : desc;
}

// portLabel localizes one wiring pin's name. Takes the label rather than the
// Port so callers can pass the manifest's label or their own fallback (the port
// id) without unpacking twice.
export function portLabel(label: string, lang?: string): string {
  if (!label) return "";
  const v = vocabularyFor(lang);
  return v?.ports[label] ?? label;
}

// The params-schema surface: one resolver per kind of string, all sharing the
// same contract as portLabel — pass the English the manifest carries, get the
// reader's language back, or that same English when there is no translation.
// Kept as separate functions rather than one `localize(kind, s)` so a call site
// reads as what it renders, and so a missing translation in one surface can't
// be masked by a hit in another (a field titled "Status" and a dropdown option
// "Status" are different strings to a translator even when they match today).
export function fieldTitle(title: string, lang?: string): string {
  if (!title) return "";
  return vocabularyFor(lang)?.fieldTitles[title] ?? title;
}

export function fieldHelp(help: string, lang?: string): string {
  if (!help) return "";
  return vocabularyFor(lang)?.fieldHelp[help] ?? help;
}

export function enumLabel(label: string, lang?: string): string {
  if (!label) return "";
  return vocabularyFor(lang)?.enums[label] ?? label;
}

// enumOptionLabel and enumValueLabel are the ONLY two ways a dropdown value
// should reach a screen. Both wrap enumLabel above; they exist because the
// mapping was open-coded in five places (three in SchemaForm, two in
// NodeCard) and the surface that had NOT open-coded it — a read-only enum on
// a node card — spent its life printing the stored identifier. So a canvas
// node read "not_equals" while the Inspector beside it read "does not equal",
// and nothing was wrong enough anywhere to notice.
//
// The two shapes are genuinely different and mixing them is the mistake worth
// preventing: building the list of options maps by INDEX (you are walking
// schema.enum), while showing what is currently set maps by VALUE (you have a
// stored string and need its label). enumValueLabel does the lookup so no
// caller has to remember which it is holding.

export function enumOptionLabel(
  schema: { enum?: unknown[]; enumNames?: string[] } | undefined,
  i: number,
  lang?: string,
): string {
  const name = schema?.enumNames?.[i];
  return name ? enumLabel(name, lang) : String(schema?.enum?.[i] ?? "");
}

export function enumValueLabel(
  schema: { enum?: unknown[]; enumNames?: string[] } | undefined,
  value: unknown,
  lang?: string,
): string {
  const str = value === undefined || value === null ? "" : String(value);
  const opts = schema?.enum ?? [];
  const i = opts.findIndex((v) => String(v) === str);
  if (i < 0) return str;
  return enumOptionLabel(schema, i, lang);
}

export function connectionText(text: string, lang?: string): string {
  if (!text) return "";
  return vocabularyFor(lang)?.connections[text] ?? text;
}

// splitConnectionNote splits a single-secret drop's connection note into the
// two things the Apps page renders from it: the field's LABEL and an example
// value for its placeholder. "Anthropic API key (sk-ant-…)." becomes
// { label: "Anthropic API key", example: "sk-ant-…" }; a note with no
// parenthetical is all label.
//
// It lives here, beside connectionText, because the label half is localized
// and the coverage guard has to reproduce the same split to know which half
// needs a translation — two copies of this regex would drift apart, and the
// half nobody noticed would be the one that stopped matching the vocabulary.
export function splitConnectionNote(note: string): {
  label: string;
  example: string;
} {
  const paren = note.match(/^(.*?)\s*\(([^)]*)\)\s*\.?$/);
  return {
    label: (paren ? paren[1] : note.replace(/\.$/, "")).trim(),
    example: paren ? paren[2] : "",
  };
}

export function nodeStateText(text: string, lang?: string): string {
  if (!text) return "";
  return vocabularyFor(lang)?.nodeState[text] ?? text;
}

export function integrationProse(
  key: string,
  english: string,
  lang?: string,
): string {
  if (!english) return "";
  const entry = vocabularyFor(lang)?.prose[key];
  if (!entry) return english;
  return entry.en === descriptionFingerprint(english) ? entry.sv : english;
}

export function integrationName(name: string, lang?: string): string {
  if (!name) return "";
  return vocabularyFor(lang)?.appNames[name] ?? name;
}

// dropLabelIsDefault reports whether `label` is still just the drop's name —
// in the catalog's English, in any language we translate into, or the bare
// module id the editor uses before the catalog arrives. The editor asks this
// before re-deriving a node's display name, so switching language renames the
// cards a user never touched while leaving a hand-typed name alone.
export function dropLabelIsDefault(
  drop: LabelledDrop & { id?: string },
  label: string,
): boolean {
  if (!label) return true;
  if (label === drop.label || label === drop.id) return true;
  for (const v of Object.values(VOCABULARY)) {
    if (v.labels[drop.label] === label) return true;
  }
  return false;
}

// dropCategoryLabel renders the category chip. Resolution is language map →
// English map → the raw value, so an unmapped locale still reads as product
// copy instead of falling all the way through to an engine enum. A category
// the maps don't know is shown as-is, which is the right failure mode: a new
// engine category surfaces visibly instead of silently rendering blank.
export function dropCategoryLabel(category: string, lang?: string): string {
  if (!category) return "";
  const v = vocabularyFor(lang);
  return v?.categories[category] ?? EN_CATEGORIES[category] ?? category;
}

File: scripts/droptext.go
// SPDX-FileCopyrightText: 2026 Angels' Ware
// SPDX-License-Identifier: AGPL-3.0-or-later

//go:build ignore

// Prints every drop id in the catalog with the English text the translations
// are made from — its name, its wiring pins, its description, and every string
// its params_schema and its connection fields put on a screen — as JSON. The
// web build's Swedish coverage guard reads the result, so adding a drop,
// adding a port, adding a param, or rewording any of them fails that test
// until the Swedish is written or refreshed.
//
// Why the description TEXT and not just the ids: i18n/drops/descriptions.sv.ts
// records the FINGERPRINT of the English each translation was made from, and
// falls back to English when it stops matching. That fail-safe is deliberate,
// but it is also silent — a reworded paragraph quietly reverts a Swedish
// reader to English with nothing anywhere saying so. Shipping the English here
// lets the guard recompute the fingerprint and name the drops that have gone
// stale.
//
// Why the PORTS: same silence, one word at a time. A pin whose label has no
// SV_PORTS entry renders its English on the card, next to pins that read
// Swedish, and nothing fails. Shipping the labels lets the guard name the drop
// the untranslated pin belongs to instead of just the orphan string.
//
// Why the rest — the label, the subtitle, the params_schema titles, help and
// enumNames, the connection fields, the "keeps state" copy: every one of them
// is looked up the same forgiving way in dropText.ts, so all of them fail
// silently too, and they had: 207 strings needed writing or re-keying — 127
// field-help paragraphs, 25 field titles, 24 connection strings and the rest
// spread over the names, the dropdowns and the keeps-state copy. Mostly whole
// drop families added after the last translation pass, plus the fallout of a
// 'wire' → 'connect' rewording that orphaned 38 translations at once. What is
// not shipped here is not guarded, which is how that happened.
//
// Ports go through core.WithPassthrough so the list is what the canvas draws,
// not what the drop declared: the universal pass pin is a label a reader sees
// and therefore a label that needs translating.
//
//	go run ./scripts/droptext.go > web/src/i18n/drops/catalog.json
package main

import (
	"encoding/json"
	"os"
	"sort"

	"github.com/dazyflow/dazyflow/core"
	_ "github.com/dazyflow/dazyflow/drops"
	"github.com/dazyflow/dazyflow/engine"
)

type entry struct {
	Description string `json:"description"`
	// Sorted and de-duplicated: a drop with three pins reading "Text" should
	// present the translator with one string, and the file should not churn
	// because a port literal moved. Every list below is built the same way.
	Ports []string `json:"ports"`
	// The card's own words: Label is the step's name, Subtitle the action line
	// under it, Integration the app chip the palette and the Inspector show.
	Label       string `json:"label"`
	Subtitle    string `json:"subtitle,omitempty"`
	Integration string `json:"integration,omitempty"`
	// The params-schema surface, one list per kind of string, matching the
	// resolver each one goes through (fieldTitle / fieldHelp / enumLabel). A
	// field titled "Status" and a dropdown option "Status" are different
	// strings to a translator even when they match today, so they are not
	// merged here either.
	Titles    []string `json:"titles,omitempty"`
	Help      []string `json:"help,omitempty"`
	EnumNames []string `json:"enum_names,omitempty"`
	// The connection card on the app's page: field labels, the help under each
	// input, and the placeholders inside them (localized too — several are
	// prose, e.g. "usually your email address").
	Connection []string `json:"connection,omitempty"`
	// The "keeps state" chip on a node card and its reset explanation.
	NodeState []string `json:"node_state,omitempty"`
	// Secret-kind connection notes, verbatim. The Apps page splits each one
	// into a field label and an example value ("Notion integration token
	// (secret_…)") and localizes the label half, so the guard needs the whole
	// note to apply the same split. OAuth notes are absent: that card names
	// the provider, never the note.
	SecretNotes []string `json:"secret_notes,omitempty"`
}

// collect appends to a de-duplicating set.
func collect(seen map[string]bool, out []string, values ...string) []string {
	for _, v := range values {
		if v == "" || seen[v] {
			continue
		}
		seen[v] = true
		out = append(out, v)
	}
	return out
}

// walkSchema pulls every human string out of a params_schema: the title and
// description on each property, and the enumNames beside an enum. It recurses
// through the places a schema nests, because a field inside an object or a
// table's row schema is still a field somebody reads.
func walkSchema(node any, e *entry, seen map[string]map[string]bool) {
	m, ok := node.(map[string]any)
	if !ok {
		return
	}
	if s, ok := m["title"].(string); ok {
		e.Titles = collect(seen["titles"], e.Titles, s)
	}
	if s, ok := m["description"].(string); ok {
		e.Help = collect(seen["help"], e.Help, s)
	}
	if arr, ok := m["enumNames"].([]any); ok {
		for _, v := range arr {
			if s, ok := v.(string); ok {
				e.EnumNames = collect(seen["enums"], e.EnumNames, s)
			}
		}
	}
	for _, key := range []string{"properties", "definitions", "$defs", "patternProperties"} {
		if props, ok := m[key].(map[string]any); ok {
			for _, v := range props {
				walkSchema(v, e, seen)
			}
		}
	}
	for _, key := range []string{"items", "additionalProperties", "if", "then", "else"} {
		if v, ok := m[key]; ok {
			walkSchema(v, e, seen)
		}
	}
	for _, key := range []string{"oneOf", "anyOf", "allOf"} {
		if arr, ok := m[key].([]any); ok {
			for _, v := range arr {
				walkSchema(v, e, seen)
			}
		}
	}
}

func main() {
	// map[id]entry — encoding/json sorts map keys, so re-runs produce clean
	// diffs the same way docsgen's output does.
	out := map[string]entry{}
	for _, m := range engine.Default.Manifests() {
		m = core.WithPassthrough(m)
		e := entry{
			Description: m.Description,
			Label:       m.Label,
			Subtitle:    m.Subtitle,
			Integration: m.Integration,
		}
		seen := map[string]map[string]bool{
			"ports": {}, "titles": {}, "help": {}, "enums": {}, "connection": {},
			"state": {}, "notes": {},
		}
		for _, p := range append(append([]core.Port{}, m.Inputs...), m.Outputs...) {
			e.Ports = collect(seen["ports"], e.Ports, p.Label)
		}
		if len(m.ParamsSchema) > 0 {
			var root any
			if err := json.Unmarshal(m.ParamsSchema, &root); err != nil {
				panic(err)
			}
			walkSchema(root, &e, seen)
		}
		for _, f := range m.ConnectionFields {
			e.Connection = collect(seen["connection"], e.Connection, f.Label, f.Help, f.Placeholder)
			e.Connection = collect(seen["connection"], e.Connection, f.Options...)
		}
		for _, r := range m.RequiresConnections {
			if r.Kind == "secret" {
				e.SecretNotes = collect(seen["notes"], e.SecretNotes, r.Note)
			}
		}
		if m.NodeState != nil {
			e.NodeState = collect(seen["state"], e.NodeState, m.NodeState.Label, m.NodeState.ResetHint)
		}
		for _, list := range [][]string{
			e.Ports, e.Titles, e.Help, e.EnumNames, e.Connection, e.NodeState,
			e.SecretNotes,
		} {
			sort.Strings(list)
		}
		out[m.ID] = e
	}
	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")
	if err := enc.Encode(out); err != nil {
		panic(err)
	}
}
